src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -79,23 +100,40 @@ public Task<Releaser> EnterAsync(CancellationToken cancellationToken = default(C
                 return DisposedReleaserTask;
             }
 
-            return this.LockWaitingHelper(this.semaphore.WaitAsync(cancellationToken));
-        }
-
-        /// <summary>
-        /// Requests access to the lock.
-        /// </summary>
-        /// <param name="timeout">A timeout for waiting for the lock.</param>
-        /// <param name="cancellationToken">A token whose cancellation signals lost interest in the lock.</param>
-        /// <returns>A task whose result is a releaser that should be disposed to release the lock.</returns>
-        public Task<Releaser> EnterAsync(TimeSpan timeout, CancellationToken cancellationToken = default(CancellationToken))
-        {
-            if (this.disposed)
+            lock (this.syncObject)
             {
-                return DisposedReleaserTask;
-            }
+                if (this.CurrentCount > 0)
+                {
+                    this.CurrentCount--;
+                    return this.uncontestedReleaser;
+                }
+                else if (timeout == TimeSpan.Zero)
+                {
+                    return CanceledReleaser;
+                }
+                else
+                {
+                    var node = this.GetNode(new WaiterInfo(this, cancellationToken));
+
+                    // Careful: consider that if the token was cancelled just now (after we checked it on entry to this method)
+                    // or the timeout expires,
+                    // then this Register method may *inline* the handler we give it, reversing the apparent order of execution with respect to
+                    // the code that follows this Register call.
+                    node.Value.CancellationTokenRegistration = cancellationToken.Register(CancellationHandler, node.Value);
+                    if (timeout != Timeout.InfiniteTimeSpan)
+                    {
+                        node.Value.TimerTokenSource = new Timer(CancellationHandler, node.Value, checked((int)timeout.TotalMilliseconds), Timeout.Infinite);
+                    }
+
+                    // Only add to the queue if cancellation hasn't already happened.
+                    if (!node.Value.Trigger.Task.IsCanceled)
+                    {
+                        this.waiters.AddLast(node);
+                    }
```

Comment:
This doesn't fix the race condition (cancellation could still happen between your check and the `AddLast`) but I think what you are doing is just an optimization and the cancellation logik will just remove the node again. Might be worth to reflect in the comment though that this is just an optimization?

Reply:
Yes, if it's inlined, there's no race because it's all on the same thread. We just have to be careful to think about both orderings.
If it _isn't_ inlined, there's no danger of a race because the cancellation handler takes the same lock that we *already* hold here, so that will force the handler to *remove* after we *add* here.